### PR TITLE
Feature: Update/Edit Todo

### PR DESCRIPTION
**Pull request by ich-wtag**
**Overview**

[Task-3: Edit/Update and Cancel Todo](https://trello.com/c/O1RQGr9p/3-task-3-update-edit-todo)
This PR introduces the functionality of Edit/Updated a todo by clicking the edit button beside input box. If any user doesn't want to change the todo-title by clicking the cancel button, the previous state will be visible.

Types of Changes

- [x] Beside each todo, there is another new button named Edit.
- [x] Clicking the Edit todo button will the turn the current Todo text content into an input box.
- [x] When in Edit Mode, The text of Add button is changed from Edit to Update.
- [x] If a user clicked the edit button, but don’t want any changes, then there should be a button named cancel to go back to the previous state.

Review
Please have a close look on this PR and kindly put your valuable comments.

[edit-update-cancel.webm](https://github.com/ich-wtag/Todo-App_Vanilla_Js/assets/153156568/138b2824-2af6-4807-9dc8-5c53e0224f35)

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -21,19 +26,107 @@ const addTodoHandler = () => {
...
+        $buttonElement.innerText = "Edit";
+        paragraphElement.textContent = inputElement.value;
+
+        displayPropertySetter(inputElement, "none");
+        displayPropertySetter(cancelButton, "none");
+        displayPropertySetter(paragraphElement, "inline-block");
+
+        todo.title = inputElement.value;
+        todo.isEditing = false;
+    }
```
Comment (hhu-wtag) [inline script.js]: 1. Use early return.
2. Let's not use thread blocking calls like `alert`. Instead we can create error notification system later on as a bonus. For now, show the error in some place of the UI.
```suggestion
    if (!todo.isEditing) {
        $buttonElement.innerText = "Update";
        inputElement.value = todo.title;

        displayPropertySetter(inputElement, "inline-block");
        displayPropertySetter(cancelButton, "inline-block");
        displayPropertySetter(paragraphElement, "none");

        todo.isEditing = true;
        
        return
    }
    if (!inputElement.value) {
        showGlobalError("You can not update an empty title.");
        return;
     }

    $buttonElement.innerText = "Edit";
    paragraphElement.textContent = inputElement.value;

    displayPropertySetter(inputElement, "none");
    displayPropertySetter(cancelButton, "none");
    displayPropertySetter(paragraphElement, "inline-block");

     todo.title = inputElement.value;
     todo.isEditing = false;
```

File: utility.js [hunk 2]
```diff
@@ -5,3 +5,7 @@ export const sanitizeInput = (value) => {
 export const clearInputField = (inputElement) => {
     inputElement.value = "";
 };
+
+export const displayPropertySetter = (element, property) => {
+    element.style.display = property;
+};
```
Comment (hhu-wtag) [inline utility.js]: Setting style element is considered an anti pattern as it creates a lot of confusion and complicity later down the road.

Let's rethink this solution of adding css through class.
```js
export const addClass = (element, class) => {
   if(element.classList.contains(class){
       return
   }
   
   element.classList.add(class)
};
```

For remove class and toggle class, we can do it in line
```js
$element.classList.toggle("hide")
```
Then we can create atomic classes that will be added and removed from elements on condition.

```css
.show {
    display: block;
}

.hide {
   display: none;
}
```

**Review by hhu-wtag — APPROVED** [review]

File: utility.js (hunk 2 above)
Comment (ich-wtag) [reply utility.js]: bhai I thought about using toggle function, but in our app there is some block element. So if I don't make them inline-block the ui structure will be changed at this point. 
So in this case, we can addCase or a custom toggler function. Which one will be better?

Comment (hhu-wtag) [reply utility.js]: Don't worry about UI structure. That will be handled when we style the whole app.

Comment (ich-wtag) [reply utility.js]: okay bhai.

**Pull request merged**